src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -0,0 +1,524 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation;
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ *
+ */
+
+#include <algorithm>    // std::sort
+#include <ns3/assert.h>
+#include <ns3/log.h>
+#include <ns3/simulator.h>
+#include <ns3/pointer.h>
+#include <ns3/boolean.h>
+
+#include "periodic-dmg-wifi-scheduler.h"
+
+namespace ns3 {
+
+NS_LOG_COMPONENT_DEFINE ("PeriodicDmgWifiScheduler");
+
+NS_OBJECT_ENSURE_REGISTERED (PeriodicDmgWifiScheduler);
+
+TypeId
+PeriodicDmgWifiScheduler::GetTypeId (void)
+{
+  static TypeId tid = TypeId ("ns3::PeriodicDmgWifiScheduler")
+    .SetParent<DmgWifiScheduler> ()
+    .SetGroupName ("Wifi")
+    .AddConstructor<PeriodicDmgWifiScheduler> ();
+  return tid;
+}
+
+PeriodicDmgWifiScheduler::PeriodicDmgWifiScheduler ()
+{
+  NS_LOG_FUNCTION (this);
+}
+
+PeriodicDmgWifiScheduler::~PeriodicDmgWifiScheduler ()
+{
+  NS_LOG_FUNCTION (this);
+}
+
+void
+PeriodicDmgWifiScheduler::DoDispose ()
+{
+  NS_LOG_FUNCTION (this);
+  DmgWifiScheduler::DoDispose ();
+}
+
+void
+PeriodicDmgWifiScheduler::UpdateStartAndRemainingTime (void)
+{
+  // for the periodic scheduler, m_allocationStartTime is useless, since the addition
+  // of new SPs is consecutive 
+  NS_LOG_FUNCTION (this);
+  if (m_addtsAllocationList.empty ())
+    {
+      // no existing allocations
+      m_remainingDtiTime = m_dtiDuration.GetMicroSeconds ();
+      m_availableSlots.push_back (std::make_pair (0, m_dtiDuration.GetMicroSeconds ()));
+    }
+  else
+    {
+      // if there are existing allocations, update DTI time just for consistency
+      m_remainingDtiTime = 0;
+      for (const auto & slot: m_availableSlots)
+        {
+          m_remainingDtiTime += (slot.second - slot.first);
+        }
+    }
+}
+
+void
+PeriodicDmgWifiScheduler::AdjustExistingAllocations (AllocationFieldListI iter, uint32_t duration, bool isToAdd)
+{
+  NS_LOG_FUNCTION (this << duration << isToAdd);
+
+  // This method is called upon a DelTsRequest or after the cleanup of 
+  // non-pseudostatic allocations.
+  // In this version of the periodic scheduler, existing allocations are not shifted 
+  // to fill the created gaps but only the vector listing the available slots is updated.
+  // For this reason, the current input parameters are useless.
+
+  auto addtsListCopy = m_addtsAllocationList;
+
+  // sort the copy to simplify the process of going through the allocation list
+  sort (addtsListCopy.begin (),
```

Comment:
As I noticed while programming, `m_addtsAllocationList` could be unordered (it depends on the admission and scheduling policy). Since `m_availableSlot` **must be ordered** to work properly on each method that makes use of it, we need to go through `addtsAllocationList` by increasing `AllocationStart` time. 
I hope this is what you meant, otherwise please let me know!

Reply:
I can't figure out why `m_addtsAllocationList` could be unordered. Because in the end you are filling `m_addtsAllocationList` given the available slots in `m_availableSlot` which should be ordered right?
Maybe we should discuss this offline